### ISSUE TITLE
New version gives bad results

### ISSUE THREAD

**Issue by twinanda** [fizyr/keras-retinanet#332]
I was working with the version committed on Dec 2, 2017. And it is working just fine. I saw that there has been many updates and I am interested in one of them being the random transform for CSV generator, so I checked out the latest version on Mar 13, 2018. There are some updates in my code that needs to be implemented to accommodate the changes (e.g., calling the new function for ResNet50RetinaNet and adapting the csv_generator input parameters), but I think there should be no issue on this part. 

However, when I try to train my network, I obtain very different results. This is the result I have after an epoch of training using the OLD implementation:

> ('init_loss', 1.1354132890701294)
> Epoch 1/5
> 
>  1/10 [==>...........................] - ETA: 39s - loss: 1.1362 - regression_loss: 0.0074 - classification_loss: 1.1288
>  2/10 [=====>........................] - ETA: 24s - loss: 1.1359 - regression_loss: 0.0076 - classification_loss: 1.1283
>  3/10 [========>.....................] - ETA: 19s - loss: 1.1352 - regression_loss: 0.0078 - classification_loss: 1.1275
>  4/10 [===========>..................] - ETA: 16s - loss: 1.1334 - regression_loss: 0.0072 - classification_loss: 1.1261
>  5/10 [==============>...............] - ETA: 13s - loss: 1.1293 - regression_loss: 0.0071 - classification_loss: 1.1222
>  6/10 [=================>............] - ETA: 10s - loss: 1.1208 - regression_loss: 0.0070 - classification_loss: 1.1138
>  7/10 [====================>.........] - ETA: 8s - loss: 1.0961 - regression_loss: 0.0068 - classification_loss: 1.0893 
>  8/10 [=======================>......] - ETA: 5s - loss: 1.0307 - regression_loss: 0.0068 - classification_loss: 1.0239
>  9/10 [==========================>...] - ETA: 2s - loss: 1.2484 - regression_loss: 0.0068 - classification_loss: 1.2416
> Epoch 00001: val_loss improved from 1.13541 to 0.67020, saving model to XXX.hdf5
> 10/10 [==============================] - 32s 3s/step - loss: 1.5443 - regression_loss: 0.0068 - classification_loss: 1.5376 - val_loss: 0.6702 - val_regression_loss: 0.0065 - val_classification_loss: 0.6637

While this is the result I got using the NEW implementation:

> ('init_loss', 5.5303945541381836)
> Epoch 1/5
>  1/10 [==>...........................] - ETA: 24s - loss: 5.0914 - regression_loss: 3.9624 - classification_loss: 1.1290
>  2/10 [=====>........................] - ETA: 12s - loss: 5.0738 - regression_loss: 3.9450 - classification_loss: 1.1288
>  3/10 [========>.....................] - ETA: 8s - loss: 5.0536 - regression_loss: 3.9250 - classification_loss: 1.1286 
>  4/10 [===========>..................] - ETA: 5s - loss: 5.0273 - regression_loss: 3.8990 - classification_loss: 1.1282
>  5/10 [==============>...............] - ETA: 4s - loss: 4.9902 - regression_loss: 3.8624 - classification_loss: 1.1278
>  6/10 [=================>............] - ETA: 3s - loss: 4.9375 - regression_loss: 3.8104 - classification_loss: 1.1271
>  7/10 [====================>.........] - ETA: 2s - loss: 4.8679 - regression_loss: 3.7419 - classification_loss: 1.1260
>  8/10 [=======================>......] - ETA: 1s - loss: 4.7975 - regression_loss: 3.6731 - classification_loss: 1.1244
>  9/10 [==========================>...] - ETA: 0s - loss: 4.7500 - regression_loss: 3.6279 - classification_loss: 1.1221
> Epoch 00001: val_loss did not improve
> 10/10 [==============================] - 6s 649ms/step - loss: 4.7202 - regression_loss: 3.6006 - classification_loss: 1.1195 - val_loss: 5.6409 - val_regression_loss: 4.5275 - val_classification_loss: 1.1134

The old implementation does detect the objects in my dataset, while the new implementation does not perform the detection correctly. I let the new implementation train for some time and also tried varying the learning rate, but it still did not converge and did not work. Because of the big difference in init_loss, I suspected that the weights might not be automatically initialized to the pre-trained resnet50, so I explicitly make the weights loaded to the model by using download_imagenet function in my code. But, still to no avail.

Anybody has an idea what could be the cause of this?

**Comment by hgaiser** [comment]
The regression loss has also changed significantly since then, the values are generally much higher now. For comparison, the regression loss for the pretrained COCO model is around 1.0 I believe. So I suppose that explains why your initial loss is higher, which is not something to worry about. However after your training I hardly see any change in the loss values. What is the code you're using to train with? Can you try the unaltered `retinanet-train` script to see if that helps? Also, can you try to visualize your dataset using `debug.py`?

**Comment by twinanda** [comment]
@hgaiser I might have missed it, but is there a documentation of how this regression loss has changed from the previous implementation? I am training with my own script since I need to convert the annotations that I have to the format accepted by the CSVGenerator. I will try your suggestions and get back to you asap. Cheers!

**Comment by hgaiser** [comment]
Pfoe I think the differences were that it can now handle batch_size > 1 and the normalization value is different (which has most impact in your case). There's no documentation other than the git history.

**Comment by twinanda** [comment]
@hgaiser Okay, I will try to trace the difference in the git history. Anyway, I tried with the bin/train.py script this is the output of the first 5 epochs: 

> Epoch 1/50
> 10/10 [==============================] - 8s 815ms/step - loss: 5.3916 - regression_loss: 4.2594 - classification_loss: 1.1322
> 
> Epoch 00001: saving model to ./snapshots/resnet50_csv_01.h5
> ('bottle', '0.0000')
> mAP: 0.0000
> Epoch 2/50
> 10/10 [==============================] - 3s 315ms/step - loss: 5.3897 - regression_loss: 4.2585 - classification_loss: 1.1312
> 
> Epoch 00002: saving model to ./snapshots/resnet50_csv_02.h5
> ('bottle', '0.0000')
> mAP: 0.0000
> Epoch 3/50
> 10/10 [==============================] - 3s 327ms/step - loss: 5.3880 - regression_loss: 4.2570 - classification_loss: 1.1310
> 
> Epoch 00003: saving model to ./snapshots/resnet50_csv_03.h5
> ('bottle', '0.0000')
> mAP: 0.0000
> Epoch 4/50
> 10/10 [==============================] - 3s 317ms/step - loss: 5.3059 - regression_loss: 4.1759 - classification_loss: 1.1301
> 
> Epoch 00004: saving model to ./snapshots/resnet50_csv_04.h5
> ('bottle', '0.0000')
> mAP: 0.0000
> Epoch 5/50
> 10/10 [==============================] - 3s 328ms/step - loss: 5.2747 - regression_loss: 4.1453 - classification_loss: 1.1293

It did however improve at the end of training:

> Epoch 50/50
> 10/10 [==============================] - 3s 315ms/step - loss: 0.1344 - regression_loss: 0.0542 - classification_loss: 0.0803
> 
> Epoch 00050: saving model to ./snapshots/resnet50_csv_50.h5
> ('bottle', '0.2000')
> mAP: 0.2000

But it still does not detect the bottles in my dataset, while the previous implentation did. I wonder if this is due to the fact that I have a small number of training data (training, test, and validation subsets have 4 images each). Would this small training data have any effects on the new regression loss computation? Note that it is working well (detects the bottles in unseen images) on previous implementation. 

I tried to visualize my dataset with debug.py. All the bounding boxes are correct.

**Comment by hgaiser** [comment]
Could you just for a sanity check compute the mAP on the training set? Those losses are quite low so I expect a near perfect score there.

**Comment by hgaiser** [comment]
Any update on this ?
